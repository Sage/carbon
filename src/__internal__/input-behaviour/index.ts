export { InputBehaviour, InputContext } from "./input-behaviour.component";
export {
  InputGroupBehaviour,
  InputGroupContext,
} from "./input-group-behaviour.component";
export type { InputContextProps } from "./useInputBehaviour";
