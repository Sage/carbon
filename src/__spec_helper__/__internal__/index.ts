import { configure } from "@testing-library/react";
import { enableFetchMocks } from "jest-fetch-mock";
import { setupMatchMediaMock } from "../mock-match-media";
import setupResizeObserverMock from "../mock-resize-observer";
import setupScrollToMock from "../mock-element-scrollto";
import "@testing-library/jest-dom";

enableFetchMocks();
setupResizeObserverMock();
setupMatchMediaMock();
setupScrollToMock();

configure({
  testIdAttribute:
    "data-role" /** Configure React Testing Library *ByTestId queries to use data-role tag */,
});
