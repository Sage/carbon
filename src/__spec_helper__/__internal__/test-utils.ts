/* eslint-disable jest/no-conditional-expect */
/* eslint-disable jest/no-identical-title */
/* eslint-disable jest/no-export */
import { render } from "@testing-library/react";
import { sprintf } from "sprintf-js";
import {
  LayoutProps,
  FlexboxProps,
  BackgroundProps,
  PositionProps,
  GridProps,
} from "styled-system";

import { space } from "../../style/themes/base/base-theme.config";
import { mockMatchMedia } from "../mock-match-media";

const isUpper = (char: string) => char.toUpperCase() === char;
const humpToDash = (acc: string, char: string) =>
  `${acc}${isUpper(char) ? `-${char.toLowerCase()}` : char}`;

const toCSSCase = (str: string) => {
  return str.split("").reduce(humpToDash, "");
};

const assertStyleMatch = (
  styleSpec: { [key: string]: string | number | undefined },
  component: HTMLElement,
  opts?: jest.Options,
) => {
  Object.entries(styleSpec).forEach(([attr, value]) => {
    expect(component).toHaveStyleRule(toCSSCase(attr), value, opts);
  });
};

const makeArrayKeys = (n: number) => [...Array(n).keys()];

const marginProps = [
  ["m", "margin"],
  ["ml", "marginLeft"],
  ["mr", "marginRight"],
  ["mt", "marginTop"],
  ["mb", "marginBottom"],
  ["mx", "marginLeft"],
  ["mx", "marginRight"],
  ["my", "marginTop"],
  ["my", "marginBottom"],
] as const;

type MarginProps = {
  [K in (typeof marginProps)[number][0]]?: string | number;
};

const paddingProps = [
  ["p", "padding"],
  ["pl", "paddingLeft"],
  ["pr", "paddingRight"],
  ["pt", "paddingTop"],
  ["pb", "paddingBottom"],
  ["px", "paddingLeft"],
  ["px", "paddingRight"],
  ["py", "paddingTop"],
  ["py", "paddingBottom"],
] as const;

type PaddingProps = {
  [K in (typeof paddingProps)[number][0]]?: string | number;
};

const colorProps = [
  ["color", "color", "#CCCCCC"],
  ["bg", "background-color", "#FFFFFF"],
  ["opacity", "opacity", "0.5"],
] as const;

const widthProps = ["width", "width", "200px"] as const;

const heightProps = ["height", "height", "200px"] as const;

const layoutProps = [
  widthProps,
  heightProps,
  ["minWidth", "min-width", "120px"],
  ["maxWidth", "max-width", "120px"],
  ["minHeight", "min-height", "120px"],
  ["maxHeight", "max-height", "120px"],
  ["size", "width", "120px"],
  ["size", "height", "120px"],
  ["display", "display", "inline-block"],
  ["verticalAlign", "vertical-align", "baseline"],
  ["overflow", "overflow", "hidden"],
  ["overflowX", "overflow-x", "hidden"],
  ["overflowY", "overflow-y", "hidden"],
] as const;

const flexBoxProps = [
  ["alignItems", "alignItems", "center"],
  ["alignContent", "alignContent", "center"],
  ["justifyItems", "justifyItems", "center"],
  ["justifyContent", "justifyContent", "center"],
  ["flexWrap", "flexWrap", "wrap"],
  ["flexDirection", "flexDirection", "row-reverse"],
  ["flex", "flex", "1"],
  ["flexGrow", "flexGrow", "1"],
  ["flexShrink", "flexShrink", "1"],
  ["flexBasis", "flexBasis", "100px"],
  ["justifySelf", "justifySelf", "center"],
  ["alignSelf", "alignSelf", "center"],
  ["order", "order", "1"],
] as const;

const gridProps = [
  ["gridColumn", "gridColumn", "1 / span 2"],
  ["gridRow", "gridRow", "1 / span 2"],
  ["gridArea", "gridArea", "myArea"],
  ["gridAutoFlow", "gridAutoFlow", "column"],
  ["gridAutoRows", "gridAutoRows", "150px"],
  ["gridAutoColumns", "gridAutoColumns", "50px"],
  ["gridTemplateRows", "gridTemplateRows", "100px 300px"],
  ["gridTemplateColumns", "gridTemplateColumns", "auto auto auto auto"],
  ["gridTemplateAreas", "gridTemplateAreas", "myArea myArea . . ."],
] as const;

const backgroundProps = [
  [
    "background",
    "background",
    "lightblue url('foo.jpg') no-repeat fixed center",
  ],
  ["backgroundImage", "background-image", "url(foo.jpg)"],
  ["backgroundSize", "background-size", "center"],
  ["backgroundRepeat", "background-repeat", "no-repeat"],
] as const;

const positionProps = [
  ["top", "0px"],
  ["bottom", "0px"],
  ["right", "0px"],
  ["left", "0px"],
  ["position", "fixed"],
  ["position", "sticky"],
  ["position", "absolute"],
  ["position", "static"],
  ["position", "relative"],
] as const;

export const getDefaultValue = (value?: string | number) => {
  const spaceArrayLength = space.length - 1;
  if (value === undefined) return value;
  const parsedValue = +value;
  if (typeof value === "string" && parsedValue > spaceArrayLength) {
    return `${value}px`;
  }
  if (parsedValue <= spaceArrayLength) {
    return space[parsedValue];
  }
  if (parsedValue > spaceArrayLength) {
    return `${parsedValue * 8}px`;
  }
  return value;
};

const testStyledSystemGrid = (
  component: (gridProperties?: GridProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(gridProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component(props));

        assertStyleMatch({ [propName]: value }, elementQuery());
      });
    },
  );
};

const testStyledSystemBackground = (
  component: (backgroundProperties?: BackgroundProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(backgroundProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));

        assertStyleMatch({ [styledSystemProp]: value }, elementQuery());
      });
    },
  );
};

const testStyledSystemPosition = (
  component: (positionProperties?: PositionProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(positionProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, value) => {
      it(`then ${styledSystemProp} should have been set correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));

        assertStyleMatch({ [styledSystemProp]: value }, elementQuery());
      });
    },
  );
};

// this util will catch that a console output occurred without polluting the output when running the unit tests
const expectConsoleOutput = (
  message: string,
  type: "warn" | "error" = "error",
) => {
  if (!message) {
    throw new Error(`no ${type} message provided`);
  }

  expect.assertions(1);

  const consoleType = global.console[type];
  let consoleArgs: string[];

  jest.spyOn(global.console, type).mockImplementation((...args) => {
    if (!args.length) return;

    const msg = args.join(" ");
    const params = args.slice(1, args.length);

    if (sprintf(msg, ...params).includes(message)) {
      consoleArgs = args;
      return;
    }

    consoleType(...args);
  });

  return () => {
    if (consoleArgs) {
      // eslint-disable-next-line no-console
      expect(console[type]).toHaveBeenCalledWith(...consoleArgs);
    }

    global.console[type] = consoleType;
  };
};

const testStyledSystemMargin = (
  component: (spacingProps?: MarginProps) => JSX.Element,
  elementQuery: () => HTMLElement,
  defaults?: MarginProps,
  assertOpts?: jest.Options,
) => {
  describe("default props", () => {
    let StyleElement: HTMLElement;

    beforeAll(() => {
      render(component({ ...defaults }));
      StyleElement = elementQuery();
    });

    it("should set the correct margins", () => {
      let margin;
      let marginLeft;
      let marginRight;
      let marginTop;
      let marginBottom;

      if (defaults) {
        margin = getDefaultValue(defaults.m);
        marginLeft = getDefaultValue(defaults.ml || defaults.mx);
        marginRight = getDefaultValue(defaults.mr || defaults.mx);
        marginTop = getDefaultValue(defaults.mt || defaults.my);
        marginBottom = getDefaultValue(defaults.mb || defaults.my);

        assertStyleMatch(
          {
            margin,
            marginLeft,
            marginRight,
            marginTop,
            marginBottom,
          },
          StyleElement,
          assertOpts,
        );
      } else {
        expect(StyleElement).not.toHaveStyleRule("marginLeft");
        expect(StyleElement).not.toHaveStyleRule("marginRight");
        expect(StyleElement).not.toHaveStyleRule("marginTop");
        expect(StyleElement).not.toHaveStyleRule("marginBottom");
        expect(StyleElement).not.toHaveStyleRule("margin");
      }
    });
  });

  describe.each(marginProps)(
    'when a custom spacing is specified using the "%s" styled system props',
    (styledSystemProp, propName) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: 2 };
        render(component({ ...props }));

        assertStyleMatch(
          { [propName]: "var(--spacing200)" },
          elementQuery(),
          assertOpts,
        );
      });
    },
  );
};

const testStyledSystemPadding = (
  component: (spacingProps?: PaddingProps) => JSX.Element,
  elementQuery: () => HTMLElement,
  defaults?: PaddingProps,
  assertOpts?: jest.Options,
) => {
  describe("default props", () => {
    let StyleElement: HTMLElement;

    beforeAll(() => {
      render(component({ ...defaults }));
      StyleElement = elementQuery();
    });

    it("should set the correct paddings", () => {
      let padding;
      let paddingLeft;
      let paddingRight;
      let paddingTop;
      let paddingBottom;

      if (defaults) {
        padding = getDefaultValue(defaults.p);
        paddingLeft = getDefaultValue(defaults.pl || defaults.px);
        paddingRight = getDefaultValue(defaults.pr || defaults.px);
        paddingTop = getDefaultValue(defaults.pt || defaults.py);
        paddingBottom = getDefaultValue(defaults.pb || defaults.py);

        assertStyleMatch(
          {
            padding,
            paddingLeft,
            paddingRight,
            paddingTop,
            paddingBottom,
          },
          StyleElement,
          assertOpts,
        );
      } else {
        expect(StyleElement).not.toHaveStyleRule("paddingLeft");
        expect(StyleElement).not.toHaveStyleRule("paddingRight");
        expect(StyleElement).not.toHaveStyleRule("paddingTop");
        expect(StyleElement).not.toHaveStyleRule("paddingBottom");
        expect(StyleElement).not.toHaveStyleRule("padding");
      }
    });
  });

  describe.each(paddingProps)(
    'when a custom spacing is specified using the "%s" styled system props',
    (styledSystemProp, propName) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: 2 };
        render(component({ ...props }));

        assertStyleMatch(
          { [propName]: "var(--spacing200)" },
          elementQuery(),
          assertOpts,
        );
      });
    },
  );
};

const testStyledSystemSpacing = (
  component: (spacingProps?: MarginProps | PaddingProps) => JSX.Element,
  elementQuery: () => HTMLElement,
  defaults?: MarginProps | PaddingProps,
  assertOpts?: jest.Options,
) => {
  testStyledSystemMargin(
    component,
    elementQuery,
    defaults as MarginProps,
    assertOpts,
  );
  testStyledSystemPadding(
    component,
    elementQuery,
    defaults as PaddingProps,
    assertOpts,
  );
};

const testStyledSystemColor = (
  // https://stackoverflow.com/questions/53711454/styled-system-props-typing-with-typescript
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  component: (colorProperties?: any) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(colorProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));
        const StyleElement = elementQuery();
        // Some props need to have camelcase so used toHaveStyleRule rather than assertStyleMatch
        expect(StyleElement).toHaveStyleRule(propName, value);
      });
    },
  );
};

const testStyledSystemWidth = (
  component: (widthProperties?: { width: string }) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe("when a width prop is specified using styled system props", () => {
    it("should set the width styling correctly", () => {
      const [styledSystemProp, propName, value] = widthProps;
      const props = { [styledSystemProp]: value };
      render(component({ ...props }));
      const StyleElement = elementQuery();
      expect(StyleElement).toHaveStyleRule(propName, value);
    });
  });
};

const testStyledSystemHeight = (
  component: (heightProperties?: { height: string }) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe("when a height prop is specified using styled system props", () => {
    it("should set the height styling correctly", () => {
      const [styledSystemProp, propName, value] = heightProps;
      const props = { [styledSystemProp]: value };
      render(component({ ...props }));
      const StyleElement = elementQuery();
      expect(StyleElement).toHaveStyleRule(propName, value);
    });
  });
};

const testStyledSystemLayout = (
  component: (layoutProperties?: LayoutProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(layoutProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));
        const StyleElement = elementQuery();
        // Some props need to have camelcase so used toHaveStyleRule rather than assertStyleMatch
        expect(StyleElement).toHaveStyleRule(propName, value);
      });
    },
  );
};

const testStyledSystemFlexBox = (
  component: (flexboxProperties?: FlexboxProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(flexBoxProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component(props));

        assertStyleMatch({ [propName]: value }, elementQuery());
      });
    },
  );
};

export {
  assertStyleMatch,
  toCSSCase,
  makeArrayKeys,
  mockMatchMedia,
  testStyledSystemColor,
  testStyledSystemLayout,
  testStyledSystemFlexBox,
  testStyledSystemGrid,
  testStyledSystemBackground,
  testStyledSystemPosition,
  expectConsoleOutput,
  testStyledSystemSpacing,
  testStyledSystemMargin,
  testStyledSystemPadding,
  testStyledSystemWidth,
  testStyledSystemHeight,
};
