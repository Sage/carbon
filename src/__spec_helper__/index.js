import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { setup } from './mock-match-media';

setup();
Enzyme.configure({ adapter: new Adapter() });
