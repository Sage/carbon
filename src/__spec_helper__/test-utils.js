import { mount } from 'enzyme';

import { carbonThemeList } from '../style/themes';
import { mockMatchMedia } from './mock-match-media';

const isUpper = char => char.toUpperCase() === char;
const humpToDash = (acc, char) => `${acc}${isUpper(char) ? `-${char.toLowerCase()}` : char}`;

const toCSSCase = (str) => {
  return str.split('').reduce(humpToDash, '');
};

const assertStyleMatch = (styleSpec, component, opts) => {
  Object.entries(styleSpec).forEach(([attr, value]) => {
    expect(component).toHaveStyleRule(toCSSCase(attr), value, opts);
  });
};

const makeArrayKeys = n => [...Array(n).keys()];

const dispatchKeyPress = (code) => {
  const ev = new KeyboardEvent('keydown', { which: code });
  document.dispatchEvent(ev);
};

const keyMap = {
  UpArrow: '38',
  DownArrow: '40',
  RightArrow: '39',
  LeftArrow: '37',
  Enter: '13',
  Tab: '9',
  Space: '32',
  Escape: '27',
  End: '35',
  Home: '36',
  D: '68',
  E: '69',
  P: '80',
  Z: '90',
  1: '49'
};

const repeat = action => (n = 1) => makeArrayKeys(n).forEach(() => action());

const keyboard = Object.keys(keyMap).reduce((acc, key) => {
  acc[`press${key}`] = () => repeat(dispatchKeyPress(keyMap[key]));
  return acc;
}, {});

// Build an object of Enzyme simulate helpers
// e.g. simulate.keydown.pressTab(target, { shiftKey: true })
// e.g. simulate.keydown.pressEscape(target)
const keydown = Object.keys(keyMap).reduce((acc, key) => {
  acc[`press${key}`] = (target, { shiftKey } = { shiftKey: false }) => {
    target.simulate('keydown', { shiftKey, key, which: parseInt(keyMap[key], 0) });
  };
  return acc;
}, {});

const simulate = {
  keydown
};

const listFrom = wrapper => wrapper.find('ul');
const childrenFrom = node => node.children();

const hoverList = wrapper => (item) => {
  childrenFrom(listFrom(wrapper)).at(item).simulate('mouseover');
};

const simulateEvent = eventName => wrapper => wrapper.simulate(eventName);
const click = simulateEvent('click');

const selectedItemOf = wrapper => wrapper.state().selectedItem;
const isUnique = (val, index, self) => self.indexOf(val) === index;
const isSelectableGiven = nonSelectables => i => !nonSelectables.includes(i);

const selectedItemReducer = method => wrapper => (acc, i) => {
  method(wrapper)(i);
  return [...acc, selectedItemOf(wrapper)];
};

const arraysEqual = (arr1, arr2) => arr1.sort().join(',') === arr2.sort().join(',');

const assertCorrectTraversal = method => expect => ({ num, nonSelectables = [] }) => (wrapper) => {
  const array = makeArrayKeys(num);
  const validIndexes = array.filter(isSelectableGiven(nonSelectables));


  const selectedItem = selectedItemOf(wrapper);
  const indexesThatWereSelected = array
    .reduce(selectedItemReducer(method)(wrapper), [selectedItem])
    .filter(isUnique);
  expect(arraysEqual(validIndexes, indexesThatWereSelected)).toBeTruthy();
};

const assertKeyboardTraversal = assertCorrectTraversal(() => keyboard.pressDownArrow)(expect);
const assertHoverTraversal = assertCorrectTraversal(wrapper => hoverList(wrapper))(expect);

const carbonThemesJestTable = carbonThemeList.map(theme => [theme.name, theme]);

const spacingProps = [
  ['m', 'margin'],
  ['ml', 'marginLeft'],
  ['mr', 'marginRight'],
  ['mt', 'marginTop'],
  ['mb', 'marginBottom'],
  ['mx', 'marginLeft'],
  ['mx', 'marginRight'],
  ['my', 'marginTop'],
  ['my', 'marginBottom'],
  ['p', 'padding'],
  ['pl', 'paddingLeft'],
  ['pr', 'paddingRight'],
  ['pt', 'paddingTop'],
  ['pb', 'paddingBottom'],
  ['px', 'paddingLeft'],
  ['px', 'paddingRight'],
  ['py', 'paddingTop'],
  ['py', 'paddingBottom']
];

const getDefaultValue = (value) => {
  if (typeof value === 'number') {
    return `${value * 8}px`;
  }
  return value;
};

const testStyledSystemSpacing = (component, defaults, styleContainer) => {
  describe('default props', () => {
    const wrapper = mount(component());
    const StyleElement = styleContainer ? styleContainer(wrapper) : wrapper;

    it('should set the correct margins', () => {
      let margin;
      let marginLeft;
      let marginRight;
      let marginTop;
      let marginBottom;

      if (defaults) {
        margin = getDefaultValue(defaults.m || undefined);
        marginLeft = getDefaultValue(defaults.ml || defaults.mx || undefined);
        marginRight = getDefaultValue(defaults.mr || defaults.mx || undefined);
        marginTop = getDefaultValue(defaults.mt || defaults.my || undefined);
        marginBottom = getDefaultValue(defaults.mb || defaults.my || undefined);

        expect(assertStyleMatch(
          {
            margin,
            marginLeft,
            marginRight,
            marginTop,
            marginBottom
          },
          StyleElement
        ));
      } else {
        expect(StyleElement).not.toHaveStyleRule('marginLeft');
        expect(StyleElement).not.toHaveStyleRule('marginRight');
        expect(StyleElement).not.toHaveStyleRule('marginTop');
        expect(StyleElement).not.toHaveStyleRule('marginBottom');
        expect(StyleElement).not.toHaveStyleRule('margin');
      }
    });

    it('should set the correct paddings', () => {
      let padding;
      let paddingLeft;
      let paddingRight;
      let paddingTop;
      let paddingBottom;

      if (defaults) {
        padding = getDefaultValue(defaults.p || undefined);
        paddingLeft = getDefaultValue(defaults.pl || defaults.px || undefined);
        paddingRight = getDefaultValue(defaults.pr || defaults.px || undefined);
        paddingTop = getDefaultValue(defaults.pt || defaults.py || undefined);
        paddingBottom = getDefaultValue(defaults.pb || defaults.py || undefined);

        expect(assertStyleMatch(
          {
            padding,
            paddingLeft,
            paddingRight,
            paddingTop,
            paddingBottom
          },
          StyleElement
        ));
      } else {
        expect(StyleElement).not.toHaveStyleRule('paddingLeft');
        expect(StyleElement).not.toHaveStyleRule('paddingRight');
        expect(StyleElement).not.toHaveStyleRule('paddingTop');
        expect(StyleElement).not.toHaveStyleRule('paddingBottom');
        expect(StyleElement).not.toHaveStyleRule('padding');
      }
    });
  });

  describe.each(spacingProps)('when a custom spacing is specified using the "%s" styled system props',
    (styledSystemProp, propName) => {
      it(`then that ${propName} should have been set correctly`, () => {
        let wrapper = mount(component());

        const props = { [styledSystemProp]: 2 };
        wrapper = mount(component({ ...props }));

        expect(assertStyleMatch(
          { [propName]: '16px' },
          styleContainer ? styleContainer(wrapper) : wrapper
        ));
      });
    });
};

export {
  assertStyleMatch,
  toCSSCase,
  hoverList,
  selectedItemOf,
  childrenFrom,
  makeArrayKeys,
  keyboard,
  assertKeyboardTraversal,
  assertHoverTraversal,
  listFrom,
  click,
  simulate,
  carbonThemesJestTable,
  mockMatchMedia,
  testStyledSystemSpacing
};
