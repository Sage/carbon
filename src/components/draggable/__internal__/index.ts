export { DragDropProvider, useDragDropContext } from "./drag-drop-provider";
