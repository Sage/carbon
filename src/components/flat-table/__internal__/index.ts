export { default } from "./use-calculate-sticky-cells";
