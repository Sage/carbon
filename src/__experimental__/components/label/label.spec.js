import React from 'react';
import { act } from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme';
import TestRenderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';

import Help from '../../../components/help';
import Label from './label.component';
import StyledLabel, { StyledLabelContainer } from './label.style';
import { assertStyleMatch } from '../../../__spec_helper__/test-utils';
import ValidationIcon from '../../../components/validations/validation-icon.component';
import baseTheme from '../../../style/themes/base';
import mintTheme from '../../../style/themes/mint';
import IconWrapperStyle from './icon-wrapper.style';
import { InputContext, InputGroupContext } from '../../../__internal__/input-behaviour';

const validationTypes = ['error', 'warning', 'info'];

describe('Label', () => {
  it('renders the label', () => {
    expect(render({}, TestRenderer.create)).toMatchSnapshot();
  });

  describe('context handlers', () => {
    let contextOnMouseEnter;
    let groupContextOnMouseEnter;

    let contextOnMouseLeave;
    let groupContextOnMouseLeave;

    let wrapper;

    beforeEach(() => {
      contextOnMouseEnter = jest.fn();
      groupContextOnMouseEnter = jest.fn();

      contextOnMouseLeave = jest.fn();
      groupContextOnMouseLeave = jest.fn();

      wrapper = renderWithContext({},
        {
          onMouseEnter: groupContextOnMouseEnter,
          onMouseLeave: groupContextOnMouseLeave
        },
        {
          onMouseEnter: contextOnMouseEnter,
          onMouseLeave: contextOnMouseLeave
        },);
    });


    it('triggers onMouseEnter callbacks passed from context', () => {
      wrapper.find(StyledLabel).props().onMouseEnter();
      expect(contextOnMouseEnter).toHaveBeenCalled();
      expect(groupContextOnMouseEnter).toHaveBeenCalled();
    });

    it('triggers onMouseLeave callbacks passed from context', () => {
      wrapper.find(StyledLabel).props().onMouseLeave();
      expect(contextOnMouseLeave).toHaveBeenCalled();
      expect(groupContextOnMouseLeave).toHaveBeenCalled();
    });

    it('does nothing if onMouseEnter or onMouseLeave callbacks are not provided', () => {
      wrapper = render();
      const inputProps = wrapper.find(StyledLabel).props();
      inputProps.onMouseEnter();
      inputProps.onMouseLeave();
    });
  });

  describe('when initiated with the help prop', () => {
    it('contains Help component with the content specified in that prop', () => {
      const wrapper = render({ help: 'Help me!' }, shallow);
      expect(wrapper.find(Help).contains('Help me!')).toBe(true);
    });
  });

  describe('when inline', () => {
    it('applies styling for an inline label', () => {
      const wrapper = render({ inline: true });
      assertStyleMatch({
        boxSizing: 'border-box',
        marginBottom: '0',
        justifyContent: 'flex-end',
        width: '30%'
      }, wrapper.find(StyledLabelContainer));
    });

    it('applies correct styling for inline label with 0 width', () => {
      const wrapper = render({ inline: true, width: 0 });
      assertStyleMatch({
        width: '30%'
      }, wrapper.find(StyledLabelContainer));
    });

    it('applies styling for an inline "optional" label', () => {
      const wrapper = render({
        inline: true,
        childOfForm: true,
        optional: true,
        theme: mintTheme
      });

      assertStyleMatch({
        content: "'(optional)'",
        fontWeight: '350',
        marginLeft: '4px'
      }, wrapper.find(StyledLabelContainer),
      { modifier: '::after' });
    });

    it('applies styling when pr prop set', () => {
      assertStyleMatch({
        paddingRight: '16px'
      }, render({ inline: true, pr: 2 }, TestRenderer.create).toJSON());
    });
  });

  describe('when left aligned', () => {
    it('aligns the label correctly', () => {
      const wrapper = render({ inline: true, align: 'left' });
      assertStyleMatch({
        justifyContent: 'flex-start'
      }, wrapper.find(StyledLabelContainer));
    });
  });

  describe('when disabled', () => {
    it('applies disabled color', () => {
      const wrapper = render({ disabled: true });

      assertStyleMatch({
        color: baseTheme.disabled.disabled
      }, wrapper.find(StyledLabel));
    });
  });

  describe('with readonly', () => {
    it('applies disabled color', () => {
      const wrapper = render({ error: true, readOnly: true });
      assertStyleMatch({
        color: baseTheme.text.color
      }, wrapper.find(StyledLabel));
    });
  });

  describe('with disabled', () => {
    it('applies disabled color', () => {
      const wrapper = render({ error: true, disabled: true });

      assertStyleMatch({
        color: baseTheme.disabled.disabled
      }, wrapper.find(StyledLabel));
    });
  });

  describe('when the help icon is focused', () => {
    it('then the IconWrapper outline should have the expected value', () => {
      const wrapper = renderWithTheme({ help: 'help message' }, baseTheme).find(IconWrapperStyle);
      wrapper.simulate('focus');

      assertStyleMatch({
        outline: `2px solid ${baseTheme.colors.focus}`
      },
      wrapper, { modifier: ':focus' });
    });
  });

  describe('when attached to child of form', () => {
    describe('when inline', () => {
      it('applies styling for an inline label', () => {
        assertStyleMatch({
          marginLeft: '12px'
        }, render({ childOfForm: true, inline: true, align: 'right' }, TestRenderer.create).toJSON());
      });
    });

    describe('when not inline', () => {
      it('applies styling for label', () => {
        assertStyleMatch({
          marginBottom: '12px'
        }, render({ childOfForm: true }, TestRenderer.create).toJSON());
      });
    });

    describe('when IconWrapperStyle', () => {
      let wrapper;

      beforeEach(() => {
        wrapper = render({
          help: 'Message'
        }, mount);
      });

      describe('will run `onFocus` event', () => {
        it('should change `isFocused` to be true', () => {
          act(() => {
            wrapper.find(IconWrapperStyle).simulate('focus');
          });
          wrapper.update();

          expect(wrapper.find(Help).props().isFocused).toBe(true);
        });
      });

      describe('will run `onBlur` event', () => {
        it('should change `isFocused` to be false', () => {
          act(() => {
            wrapper.find(IconWrapperStyle).simulate('blur');
          });

          wrapper.update();

          expect(wrapper.find(Help).props().isFocused).toBe(false);
        });
      });
    });
  });

  describe.each(validationTypes)('when %s prop is passed as string', (vType) => {
    it('show validation icon', () => {
      const wrapper = render({ [vType]: 'Message', useValidationIcon: true }, mount);
      const icon = wrapper.find(ValidationIcon);

      expect(icon.exists()).toEqual(true);
    });
  });

  describe.each(validationTypes)('when %s prop is passed as true boolean', (vType) => {
    it('do not show validation icon', () => {
      const wrapper = render({ [vType]: true, useValidationIcon: true }, mount);
      const icon = wrapper.find(ValidationIcon);

      expect(icon.exists()).toEqual(false);
    });
  });
});

function render(props, renderer = mount) {
  return renderer(
    <Label { ...props }>
      Name:
    </Label>
  );
}

function renderWithContext(props = {}, inputGroupContextValue = {}, inputContextValue = {}) {
  return mount(
    <InputGroupContext.Provider value={ inputGroupContextValue }>
      <InputContext.Provider value={ inputContextValue }>
        <Label { ...props }>
          Name:
        </Label>
      </InputContext.Provider>
    </InputGroupContext.Provider>
  );
}

function renderWithTheme(props = {}, theme, renderer = mount) {
  return renderer(
    <ThemeProvider theme={ theme }>
      <Label { ...props }>
        Name:
      </Label>
    </ThemeProvider>
  );
}
