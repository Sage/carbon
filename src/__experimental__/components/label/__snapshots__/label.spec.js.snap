// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label renders the label 1`] = `
.c0 {
  color: rgba(0,0,0,0.9);
  display: block;
  font-weight: 600;
  padding-bottom: 8px;
}

.c0 .c1,
.c0 .sc-bZQynM {
  vertical-align: middle;
}

<label
  className="c0"
  data-element="label"
  width={30}
>
  Name:
</label>
`;
